File: mobile-rn/src/services/apiClient.ts
import { authService } from './authService';
import type { ApiError } from './apiError';
import { DEFAULT_API_BASE_URL } from '../config/env';
import {
  parseRunConfig,
  parseStartRunResponse,
  parseRunStatus,
  runRequestToWire,
  type RunConfig,
  type RunRequest,
  type StartRunResponse,
  type RunStatus,
} from '../models/run';
import { parseChartPayload, type ChartPayload } from '../models/chart';
import type { PlaceOrderRequest } from '../models/order';
import {
  parseVoiceConfig,
  parseVoiceReconcileResponse,
  parseVoiceSessionStartResponse,
  type VoiceConfig,
  type VoiceReconcileResponse,
  type VoiceSessionStartResponse,
} from '../models/voice';

/**
 * REST client for the StockAgents backend, ported from the Swift `APIClient`.
 *
 * Responsibilities: a configurable base URL, bearer-token injection (from
 * `authService`), JSON decode against the typed model parsers, non-2xx → typed
 * `ApiError` (incl. 401 + 429-with-Retry-After), per-call timeouts, and a single
 * transparent token refresh on 401. SSE is handled separately by `sseClient`.
 */
class ApiClient {
  /** Mutable so a Settings screen can repoint the app at a new backend. */
  baseURL: string = DEFAULT_API_BASE_URL;

  /** Default per-request timeout. Runs start fast (the POST returns a run_id). */
  private timeoutMs = 60_000;

  // ── REST endpoints ──────────────────────────────────────────────────────
  //
  // Reads (config/chart/status) and SSE use the open legacy `/api/*` routes.
  // Run *creation* uses the authenticated `POST /api/v2/runs` (web/mobile/
  // router.py), which adds per-user auth + rate limiting in front of the same
  // run machinery — so the run it creates is still served by the legacy
  // `GET /api/runs/{id}` status + `/events` SSE endpoints. The bearer is
  // injected from `authService`; a 401 triggers one transparent refresh+retry.

  /** `GET /api/config` — public bootstrap call; drives the run form defaults. */
  async fetchConfig(): Promise<RunConfig> {
    return parseRunConfig(await this.request('GET', '/api/config'));
  }

  /**
   * `POST /api/v2/runs` — start a run (authenticated). Requires a signed-in
   * session; mirrors the legacy `/api/runs` discriminator: returns
   * `{cached:true, run}` on a 60-minute cache hit (same ticker + trade_date,
   * cache toggle on) and `{cached:false, run_id, nodes}` otherwise. Cache hits
   * intentionally bypass the per-user run quota. On 401 the request
   * transparently refreshes once; a 429 surfaces as a `rateLimited` error
   * carrying `Retry-After`.
   */
  async startRun(request: RunRequest): Promise<StartRunResponse> {
    return parseStartRunResponse(
      await this.request('POST', '/api/v2/runs', runRequestToWire(request)),
    );
  }

  /** `GET /api/runs/{run_id}` — run status for session restore. */
  async runStatus(runId: string): Promise<RunStatus> {
    return parseRunStatus(await this.request('GET', `/api/runs/${encodeURIComponent(runId)}`));
  }

  /** `GET /api/chart` — candles + indicators for the Market Analyst card. */
  async fetchChart(
    ticker: string,
    tradeDate: string,
    assetType = 'stock',
    lookback = 180,
  ): Promise<ChartPayload> {
    const qs = new URLSearchParams({
      ticker,
      trade_date: tradeDate,
      asset_type: assetType,
      lookback: String(lookback),
    }).toString();
    return parseChartPayload(await this.request('GET', `/api/chart?${qs}`));
  }

  // ── v2 mobile endpoints (auth + per-user broker). Defined for contract
  //    completeness; the trading UI is a deferred milestone (see README). ──

  /** `POST /api/v2/runs/{run_id}/orders` — submit the run's proposed order (v1). */
  async submitOrder(runId: string, edits: PlaceOrderRequest): Promise<unknown> {
    const body: Record<string, unknown> = {};
    if (edits.action != null) body.action = edits.action;
    if (edits.quantity != null) body.quantity = edits.quantity;
    if (edits.notional != null) body.notional = edits.notional;
    if (edits.orderType != null) body.order_type = edits.orderType;
    return this.request('POST', `/api/v2/runs/${encodeURIComponent(runId)}/orders`, body);
  }

  // ── Voice-conversational layer (additive). Default-off until the backend
  //    is wired with LiveKit + Deepgram + ElevenLabs creds. The capability
  //    probe drives Talk-button visibility on the agent cards; the start
  //    endpoint mints a LiveKit room JWT for the React Native LiveKit SDK. ──

  /**
   * `GET /api/voice/config` — capability probe. The Talk button on each
   * agent card stays hidden unless `ready === true`. Failures collapse to
   * `enabled = ready = false` so the rest of the run UI is unaffected when
   * the optional voice deps aren't installed.
   */
  async fetchVoiceConfig(): Promise<VoiceConfig> {
    try {
      return parseVoiceConfig(await this.request('GET', '/api/voice/config'));
    } catch {
      return {
        enabled: false,
        ready: false,
        reason: 'voice_unavailable',
        ttsProvider: null,
        sessionMaxSeconds: null,
        dailyMinutesPerUser: null,
        personasCount: null,
      };
    }
  }

  /**
   * `POST /api/voice/sessions` — mint a LiveKit room JWT for one agent in a
   * completed run. The room is single-user; concurrent agents from the same
   * run live in separate rooms (per the plan §8 default).
   */
  async startVoiceSession(runId: string, agentId: string): Promise<VoiceSessionStartResponse> {
    return parseVoiceSessionStartResponse(
      await this.request('POST', '/api/voice/sessions', {
        run_id: runId,
        agent_id: agentId,
      }),
    );
  }

  /**
   * `POST /api/voice/sessions/{id}/reconcile` — re-invoke the Portfolio
   * Manager with the user's objection as new context, without re-running
   * the full pipeline. Surfaces the (possibly flipped) decision and
   * updated rationale for the in-call composer card.
   */
  async reconcilePortfolioManager(
    sessionId: string,
    objection: string,
  ): Promise<VoiceReconcileResponse> {
    return parseVoiceReconcileResponse(
      await this.request(
        'POST',
        `/api/voice/sessions/${encodeURIComponent(sessionId)}/reconcile`,
        { objection },
      ),
    );
  }

  // ── plumbing ──────────────────────────────────────────────────────────

  /**
   * Build → send, with a single transparent token refresh on 401 (iOS plan
   * §3.4.1). The request is rebuilt after the refresh so the retry carries the
   * new bearer. If the refresh fails, the session is cleared and 401 propagates.
   */
  private async request(
    method: string,
    path: string,
    body?: Record<string, unknown>,
  ): Promise<unknown> {
    try {
      return await this.send(method, path, body);
    } catch (error) {
      if (isUnauthorized(error)) {
        const outcome = await authService.attemptRefresh();
        if (outcome.ok) {
          return this.send(method, path, body);
        }
        // Only a definitive rejection wipes the session; a transport failure
        // during refresh keeps the tokens and surfaces the (retryable) 401.
        if (outcome.reason === 'rejected') {
          await authService.signOut();
        }
        throw error;
      }
      throw error;
    }
  }

  private async send(
    method: string,
    path: string,
    body?: Record<string, unknown>,
  ): Promise<unknown> {
    const url = this.resolve(path);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    const headers: Record<string, string> = { Accept: 'application/json' };
    if (body !== undefined) headers['Content-Type'] = 'application/json';
    const authHeader = await authService.getAuthHeader();
    if (authHeader) headers.Authorization = authHeader;

    let res: Response;
    try {
      res = await fetch(url, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
    } catch (err) {
      clearTimeout(timer);
      if (controller.signal.aborted) throw timeout();
      throw transport(err instanceof Error ? err.message : String(err));
    }
    clearTimeout(timer);

    const text = await res.text();
    if (!(res.status >= 200 && res.status < 300)) {
      throw mapHttpError(res, text);
    }
    if (!text) return {};
    try {
      return JSON.parse(text);
    } catch (err) {
      throw decoding(err instanceof Error ? err.message : String(err));
    }
  }

  private resolve(path: string): string {
    return `${this.baseURL.replace(/\/$/, '')}${path.startsWith('/') ? path : `/${path}`}`;
  }
}

// MARK: - Error mapping

function mapHttpError(res: Response, body: string): ApiError {
  switch (res.status) {
    case 401:
      return { kind: 'unauthorized' };
    case 429: {
      const header = res.headers.get('Retry-After');
      const retryAfter = header != null ? Number.parseInt(header, 10) : NaN;
      return { kind: 'rateLimited', retryAfter: Number.isFinite(retryAfter) ? retryAfter : null };
    }
    default:
      return { kind: 'http', status: res.status, detail: extractDetail(body) };
  }
}

/** Pull FastAPI's `{"detail": ...}` body when present. */
function extractDetail(body: string): string | null {
  if (!body) return null;
  try {
    const obj = JSON.parse(body) as Record<string, unknown>;
    const detail = obj.detail;
    if (typeof detail === 'string') return detail;
    if (detail != null) return JSON.stringify(detail);
  } catch {
    // not JSON — fall through
  }
  return null;
}

function isUnauthorized(error: unknown): boolean {
  return Boolean(error && typeof error === 'object' && (error as ApiError).kind === 'unauthorized');
}

const timeout = (): ApiError => ({ kind: 'timeout' });
const transport = (message: string): ApiError => ({ kind: 'transport', message });
const decoding = (message: string): ApiError => ({ kind: 'decoding', message });

export const apiClient = new ApiClient();

File: ios/StockAgents/Services/APIClient.swift
import Foundation

/// REST client for the StockAgents backend.
///
/// Responsibilities (plan §3.3): configurable base URL, bearer-token injection,
/// JSON encode/decode against the typed models, non-2xx → `APIError` mapping,
/// and per-call timeouts. SSE is handled separately by `RunStreamClient`.
final class APIClient {
    /// Mutable so the Settings screen can repoint the app at a new backend.
    var baseURL: URL
    private let auth: AuthService
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(baseURL: URL, auth: AuthService, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.auth = auth
        self.session = session
    }

    // MARK: - Endpoints

    /// `GET /api/config` — public bootstrap call; drives the run form defaults.
    func fetchConfig() async throws -> RunConfig {
        try await get("/api/config")
    }

    /// `POST /api/runs` — start a run or return a 60-minute cache hit.
    func startRun(_ request: RunRequest) async throws -> StartRunResponse {
        try await post("/api/runs", body: request)
    }

    /// `GET /api/runs/{run_id}` — run status for session restore.
    func runStatus(runID: String) async throws -> RunStatus {
        try await get("/api/runs/\(runID)")
    }

    /// `GET /api/chart` — candles + indicators for the Market Analyst card.
    func fetchChart(ticker: String, tradeDate: String, assetType: String = "stock", lookback: Int = 180) async throws -> ChartPayload {
        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "ticker", value: ticker),
            URLQueryItem(name: "trade_date", value: tradeDate),
            URLQueryItem(name: "asset_type", value: assetType),
            URLQueryItem(name: "lookback", value: String(lookback)),
        ]
        let query = components.percentEncodedQuery ?? ""
        return try await get("/api/chart?\(query)")
    }

    // MARK: - v2 mobile endpoints (auth + per-user broker)
    //
    // These hit the additive `/api/v2/*` router (`web/mobile/router.py`), which
    // is gated behind `MOBILE_API_ENABLED` and requires a bearer token
    // (`current_user`). The bearer header is injected by `makeRequest` from
    // `AuthService` exactly like the legacy endpoints — so when no app token is
    // present these return 401 (mapped to `APIError.unauthorized`).

    /// `GET /api/v2/robinhood/authorize` — start the server-mediated OAuth flow
    /// (plan §5.1.3). Returns the authorization URL to open in the system browser
    /// plus an opaque server-side connect-session id.
    func robinhoodAuthorize() async throws -> RobinhoodAuthorizeResponse {
        try await get("/api/v2/robinhood/authorize")
    }

    /// `GET /api/v2/robinhood/status` — per-user connection status. Polled after
    /// the auth sheet bounces back; the server is the source of truth (plan
    /// §5.1.3 step 5).
    func robinhoodStatus() async throws -> RobinhoodStatus {
        try await get("/api/v2/robinhood/status")
    }

    /// Submit the LIVE order proposed by a finished run.
    ///
    /// The body carries only the ticket *edits* (`action/quantity/notional/
    /// order_type`); the server re-clamps to `max_order_notional`/buying power
    /// and **forces** the ticker to the run's own. The response is the
    /// `trade`-shaped payload from `_trade_event(...)`, which decodes into
    /// `TradeEvent` (the extra `type` key is ignored).
    ///
    /// Targets the per-user-safe v2 route `POST /api/v2/runs/{run_id}/orders`
    /// (`web/mobile/orders.py` → `place_user_order`), which resolves the broker
    /// via the per-user `BrokerRegistry` (never the `get_broker()` singleton),
    /// forces the ticker to the run's own, re-clamps size, and enforces the
    /// server's dry-run/live policy. Requires a bearer token (`current_user`),
    /// injected by `authorizationHeader`. The body carries only ticket edits and
    /// must never include a ticker.
    func submitOrder(runID: String, edits: PlaceOrderRequest) async throws -> TradeEvent {
        try await post("\(APIClient.runsPath)/\(runID)/orders", body: edits)
    }

    /// Base path for per-user v2 run resources (order submission lives at
    /// `\(runsPath)/{run_id}/orders`).
    static let runsPath = "/api/v2/runs"

    // MARK: - Voice-conversational endpoints (additive, default OFF server-side)

    /// `GET /api/voice/config` — capability probe driving the Talk button
    /// visibility. Returns `enabled=false` (and `ready=false`) when the
    /// server has the feature off, in which case the UI hides the button.
    func fetchVoiceConfig() async throws -> VoiceConfig {
        try await get("/api/voice/config")
    }

    /// `POST /api/voice/sessions` — mint a LiveKit room JWT for a one-on-one
    /// voice call with `agentID` on the finished run `runID`. The body is
    /// `{run_id, agent_id}`. The server gates on the agent being `done`
    /// (409 otherwise) and on the user's daily voice-minute cap (429).
    func startVoiceSession(runID: String, agentID: String) async throws -> VoiceSessionStartResponse {
        let body = VoiceSessionRequest(runID: runID, agentID: agentID)
        return try await post("/api/voice/sessions", body: body)
    }

    /// `POST /api/voice/sessions/{id}/reconcile` — Portfolio Manager only.
    /// Submits the user's objection and returns the (possibly flipped)
    /// updated rationale. Never re-runs the pipeline; one focused LLM call.
    func reconcilePortfolioManager(sessionID: String, objection: String) async throws -> VoiceReconcileResponse {
        let body = VoiceReconcileRequest(objection: objection)
        return try await post(
            "/api/voice/sessions/\(sessionID)/reconcile",
            body: body
        )
    }

    // MARK: - Verbs

    private func get<Response: Decodable>(_ path: String) async throws -> Response {
        try await perform(path: path, method: "GET", body: nil)
    }

    private func post<Body: Encodable, Response: Decodable>(_ path: String, body: Body) async throws -> Response {
        let data = try encoder.encode(body)
        return try await perform(path: path, method: "POST", body: data)
    }

    /// Build → send, with a single transparent token refresh on `401`
    /// (plan §3.4.1 "refreshes on 401"). The request is rebuilt after the
    /// refresh so the retry carries the *new* bearer header. If the refresh
    /// fails, the session is cleared (→ signed-out) and the `401` propagates.
    private func perform<Response: Decodable>(path: String, method: String, body: Data?) async throws -> Response {
        let request = try await makeRequest(path: path, method: method, body: body)
        do {
            return try await send(request)
        } catch APIError.unauthorized {
            guard await auth.attemptRefresh() else {
                await auth.signOut()
                throw APIError.unauthorized
            }
            let retry = try await makeRequest(path: path, method: method, body: body)
            return try await send(retry)
        }
    }

    // MARK: - Plumbing

    private func makeRequest(path: String, method: String, body: Data?) async throws -> URLRequest {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        // Runs are slow to *finish* but the POST returns fast with a run_id; a
        // generous timeout still protects against a wedged connection.
        request.timeoutInterval = 60
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let header = await auth.authorizationHeader {
            request.setValue(header, forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func send<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError where error.code == .cancelled {
            throw APIError.cancelled
        } catch {
            throw APIError.transport(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport("No HTTP response")
        }
        try APIClient.validate(http, data: data)

        do {
            return try decoder.decode(Response.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: error))
        }
    }

    /// Map a non-2xx status to a typed `APIError`, extracting FastAPI's
    /// `{"detail": ...}` body when present.
    static func validate(_ http: HTTPURLResponse, data: Data) throws {
        guard !(200...299).contains(http.statusCode) else { return }
        switch http.statusCode {
        case 401:
            throw APIError.unauthorized
        case 429:
            let retryAfter = http.value(forHTTPHeaderField: "Retry-After").flatMap(Int.init)
            throw APIError.rateLimited(retryAfter: retryAfter)
        default:
            throw APIError.http(status: http.statusCode, detail: Self.detail(from: data))
        }
    }

    private static func detail(from data: Data) -> String? {
        guard
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        if let detail = object["detail"] as? String { return detail }
        if let detail = object["detail"] { return String(describing: detail) }
        return nil
    }
}
